backend/routers/transcribe_v2.py

```diff
@@ -0,0 +1,368 @@
+    w_vad = webrtcvad.Vad()
+    w_vad.set_mode(1)
+
+    # Stream transcript
+    loop = asyncio.get_event_loop()
+
+    # Soft timeout, should < MODAL_TIME_OUT - 3m
+    timeout_seconds = 420  # 7m
+    started_at = time.time()
+
+    def _get_in_progress_memory(segments: List[dict]):
+
+        memory_id = redis_db.get_in_progress_memory_id(uid)
+        existing = None
+
+        if memory_id:
+            existing = memories_db.get_memory(uid, memory_id)
+            if existing and existing['status'] != 'in_progress':
+                existing = None
+
+        if not existing:
+            existing = memories_db.get_in_progress_memory(uid)
+
+        if existing:
+            print('_get_in_progress_memory existing', existing['id'])
+            memory = Memory(**existing)
+            memory.transcript_segments = TranscriptSegment.combine_segments(
+                memory.transcript_segments, [TranscriptSegment(**segment) for segment in segments]
+            )
+            redis_db.set_in_progress_memory_id(uid, memory.id)
+            return memory
+        started_at = datetime.now(timezone.utc) - timedelta(seconds=segments[0]['end'] - segments[0]['start'])
+        memory = Memory(
+            id=str(uuid.uuid4()),
+            uid=uid,
+            structured=Structured(),
+            language=language,
+            created_at=started_at,
+            started_at=started_at,
+            finished_at=datetime.now(timezone.utc),
+            transcript_segments=[TranscriptSegment(**segment) for segment in segments],
+            status=MemoryStatus.in_progress,
+        )
+        print('_get_in_progress_memory new', memory)
+        memories_db.upsert_memory(uid, memory_data=memory.dict())
+        redis_db.set_in_progress_memory_id(uid, memory.id)
+        return memory
+
+    async def memory_creation_timer():
+        try:
+            await asyncio.sleep(120)
+            await _create_memory()
+        except asyncio.CancelledError:
+            pass
+
+    memory_creation_task = None
+    segment_start = None
+
+    def stream_transcript(segments, _):
+        nonlocal websocket
+        nonlocal segment_start
+        nonlocal memory_creation_task
+
+        if not segments or len(segments) == 0:
+            return
+
+        # Align the start, end segment
+        if not segment_start:
+            segment_start = segments[0]["start"]
+
+        if memory_creation_task is not None:
+            memory_creation_task.cancel()
+        memory_creation_task = asyncio.create_task(memory_creation_timer())
+
+        for i, segment in enumerate(segments):
+            segment["start"] -= segment_start
+            segment["end"] -= segment_start
+            segments[i] = segment
+
+        asyncio.run_coroutine_threadsafe(websocket.send_json(segments), loop)
+
+        # TODO: what when transcript is large!
+        memory = _get_in_progress_memory(segments)  # can trigger race condition? increase soniox utterance?
```

Comment:
we could use redis to cache the processing memory > segments(sorted set to maximize scalability or just go with raw json for now, dont forget set ttl) if needed, flushing period to firestore > memory for more safety(rand % 3) - but i think redis is enough.

Reply:
Set smth similar above inside _get_in_progress_memory, caching uid: memory_id in redis, O1 operation, then retrieving the documet from firestore also O1 operation, from tests, it takes 300-400 ms this function, and the whole thing stream_transcript takes 500-600 ms